### ISSUE TITLE
Perf: Retrieving CLR property values is relatively expensive

### ISSUE THREAD

**Issue by habbes** [OData/AspNetCoreOData#1110]
During serialization, we retrieve values from the properties of a CLR object so that we can convert them into `ODataValue`s that we can store in an `ODataResource`.

The [`ResourceContext.GetPropertyValue()`]() calls [`TypedEdimStructuredObject.TryGetPropertyValue()`]() which does a couple of things:
- It uses reflection to create a property getter delegate instance
- It caches the getter in a `ConcurrentDictionary` for future look ups
- When called to retrieve a property value, it retrieves the getter from the cache and invokes it

It appears that both the dictionary lookup and invocation of the delegate are relatively costly

![image](https://github.com/OData/AspNetCoreOData/assets/8460169/03d7d9ff-f384-45dc-be24-0769628132b4)

We should investigate a faster way to create and invoke the getters, and faster caching strategy for the getters.

**Comment by julealgon** [comment]
@habbes thoughts on using source generators for some of this stuff? 

Also, have you guys started to contemplate the idea of making OData AOT-ready? I know this would depend on first supporting minimal APIs but just wondering here if the team is exploring this.

**Comment by habbes** [comment]
@julealgon, those are great questions. We haven't yet made plans for AOT support, but I think that's a great opportunity to explore, albeit would require a tremendous amount of work and changes. We have made some initial explorations into minimal API support (cc @ElizabethOkerio, @gathogojr).

Regarding source generators, before your question, I envisioned that source generators would work better with a different architecture (see https://github.com/OData/odata.net/issues/2798). I anticipate that it would be hard to get considerable benefits from resource generators due the current serialization's pattern reliance on ODataResources, the dynamic nature of OData responses and all the dynamic configuration is retrieved at runtime. However, after thinking about your question, I think it may be worth exploring how much we can move at compile time even with the current serialization architecture (i.e. while keeping changes local to AspNetCoreOData without changing the core libraries). I could carve out some time to experiment with source generators to see what's possible, maybe we can generate typed converters that can convert a CLR object into an OData resource more efficiently. Maybe some configuration (like how to map certain types, enum values, etc.) could be configured at compile time via attributes instead runtime dictionary-based lookups, etc.
